Release/7.1 Eventing Listening to Multiple Collections

jon-strabala opened the pull request:
Need information about Listening to Multiple Collections in Eventing and a concrete example as per 7.1.1

jon-strabala committed:
Update for Listening to Multiple Collections

jon-strabala committed:
Update for Listening to Multiple Collections

jon-strabala committed:
Update for Listening to Multiple Collections

jon-strabala committed:
Update for Listening to Multiple Collections

jon-strabala committed:
rename file

jon-strabala committed:
Update for Listening to Multiple Collections

Add navigation

jon-strabala committed:
New Eventing example multiCollectionEventing

jon-strabala committed:
Multiple Collections (link to example)

RayOffiah commented on modules/eventing/pages/eventing-Terminologies.adoc:
```diff
@@ -110,6 +110,8 @@ NOTE: You can have multiple Eventing Functions running different code listening
 
 When a source collection is deleted, all deployed (or paused) Eventing Functions associated with this source collection are undeployed.
 
+The Listen To Location as of 7.1.1 can listen to multiple collections via a wildcard of "+++*+++" for the scope and/or the collection. For these Functions if the bucket binding used by the JavaScript code also contains a wild card of "+++*+++" for the scope and/or the collection only the Advanced Keyspace Accessors will be able to read or write the Data Service (or KV). [.status]#Couchbase Server ...
```
```suggestion
The `Listen To` Location as of 7.1.1 can listen to multiple collections via a wildcard of "+++*+++" for the scope and/or the collection. 
For these Functions, if the bucket binding used by the JavaScript code also contains a wild card of "+++*+++" for the scope and/or the collection then only the Advanced Keyspace Accessors will be able to read or write the Data Service (or KV). [.status]#Couchbase Server 7.1.1#
```

RayOffiah replied:
I would be more inclined to use backticks rather than quotes for the wildcards:
```suggestion
The Listen To Location as of 7.1.1 can listen to multiple collections via a wildcard of `+++*+++` for the scope and/or the collection. 
For these Functions if the bucket binding used by the JavaScript code also contains a wild card of `+++*+++` for the scope and/or the collection only the Advanced Keyspace Accessors will be able to read or write the Data Service (or KV). [.status]#Couchbase Server 7.1.1#
```

RayOffiah commented on modules/eventing/pages/eventing-Terminologies.adoc:
```diff
@@ -186,6 +188,8 @@ An Eventing Function can have no binding, one binding, or several bindings.  The
 
 Bucket aliases allow the JavaScript in an Eventing Function to access Couchbase KV collections from the Data Service or KV. The keyspaces (bucket.scope.collection) are then accessible by the bound name as a JavaScript map in the global space of the Eventing Function. 
 
+Prior to 7.1.1 all bucket aliases had to reference a single keyspace. As of 7.1.1 an Eventing Function can listen to multiple collections via a wildcard of "+++*+++" for the scope and/or the collection. For these Functions if the bucket alias (or binding) used by the JavaScript code can also contains a wild card of "+++*+++" for the scope and/or the collection.  If bucket alias contains a wild ca ...
```
```suggestion
Prior to 7.1.1, all bucket aliases had to reference a single keyspace. 
As of 7.1.1 an Eventing Function can listen to multiple collections via a wildcard of `+++*+++` for the scope and/or the collection. 
For these Functions, the bucket alias (or binding) used by the JavaScript code can also contains a wild card of `+++*+++` for the scope and/or the collection.  
If bucket alias contains a wild card of `+++*+++` only the Advanced Keyspace Accessors will be able to read or write the Data Service (or KV). [.status]#Couchbase Server 7.1.1#
```

RayOffiah commented on modules/eventing/pages/eventing-advanced-keyspace-accessors.adoc:
```diff
@@ -23,11 +23,20 @@ These seven (7) Advanced Keyspace Accessors make it possible to utilize and leve
 For example rather than blindly relying on he Basic Keyspace Accessors for an upsert like operation ‘src_col[id_str] = some_doc’ the Advanced Accessors allow you to resolve contention (or possible contention) with JavaScript driven logic in your Eventing Function.
 
 * If the document doesn't exist you can use ‘couchbase.insert(src_col, {"id: id_str}, some_doc)’ and check the return value for success
+* If you don't care if the document exists you can use ‘couchbase.upsert(src_col, {"id: id_str}, some_doc)’ and check the return value for success
```
```suggestion
* You can use ‘couchbase.upsert(src_col, {"id: id_str}, some_doc)’ to update a document if it already exists. 
If the document doesn't exist, then a new document will be created. 
Check the return value to see if the operation was successful.
```

RayOffiah commented on modules/eventing/pages/eventing-advanced-keyspace-accessors.adoc:
```diff
@@ -23,11 +23,20 @@ These seven (7) Advanced Keyspace Accessors make it possible to utilize and leve
 
+*Functions that Listen to Multiple Collections*: may require the _meta.keyspace_ parameter [.status]#Couchbase Server 7.1.1#
+
+For Eventing Functions that listen to multiple collections via a wildcard of "+++*+++" for the scope and/or the collection an additional parameter is be required if the binding used by the Advanced Keyspace Accessor also contains a wild card of "+++*+++" for the scope and/or the collection.
```
```suggestion
For Eventing Functions that listen to multiple collections via a wildcard of `+++*+++` for the scope and/or the collection, an additional parameter is required if the binding used by the Advanced Keyspace Accessor also contains a wild card of `+++*+++` for the scope and/or the collection.
```

RayOffiah commented on modules/eventing/pages/eventing-advanced-keyspace-accessors.adoc:
```diff
@@ -23,11 +23,20 @@ These seven (7) Advanced Keyspace Accessors make it possible to utilize and leve
+
+* ‘couchbase.insert(src_col, {"id: id_str, "keyspace":{"bucket_name": "bkt01","scope_name":"scp01","collection_name":"col01"}}, some_doc)’ 
+
+The example above adds a _keyspace_ to the _meta_ object to specify win which keyspace the insert is to take place.  Note as of 7.1.1 the entry points of OnUpdate() and OnDelete() provide _meta.keyspace_ with every mutation.  For a detailed example refer to the multiCollectionEventing example in xref:eventing-examples.adoc#examples-scriptlets-advanced-accessors[Scriptlets: Advanced Accessor Handl ...
```
```suggestion
The example above adds a _keyspace_ to the _meta_ object to specify win which keyspace the insert is to take place.  
Note as of 7.1.1 the entry points of OnUpdate() and OnDelete() provide _meta.keyspace_ with every mutation.  
For a detailed example refer to the multiCollectionEventing example in xref:eventing-examples.adoc#examples-scriptlets-advanced-accessors[Scriptlets: Advanced Accessor Handlers] 
```

RayOffiah commented on modules/eventing/pages/eventing-advanced-keyspace-accessors.adoc:
```diff
@@ -133,6 +142,10 @@ The positional parameter (denoted by _meta_ in the prototype above) represents t
 
+** _meta.keyspace (type: Object)_
++
+The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. This parameter is only required if you are using a a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```
```suggestion
The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. 
This parameter is only required if you are using a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```

RayOffiah commented on modules/eventing/pages/eventing-advanced-keyspace-accessors.adoc:
```diff
@@ -255,6 +268,10 @@ The positional parameter (denoted by _meta_ in the prototype above) represents t
 
+** _meta.keyspace (type: Object)_
++
+The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. This parameter is only required if you are using a a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```
```suggestion
The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. 
This parameter is only required if you are using  a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```

RayOffiah commented on modules/eventing/pages/eventing-advanced-keyspace-accessors.adoc:
```diff
@@ -373,6 +390,10 @@ The positional parameter (denoted by _meta_ in the prototype above) represents t
 +
 The key of the document to be used for the operation. This is a mandatory parameter and must be of JavaScript _string_ type.
```
Is all this supposed to be in bold?

RayOffiah commented on modules/eventing/pages/eventing-advanced-keyspace-accessors.adoc:
```diff
@@ -373,6 +390,10 @@ The positional parameter (denoted by _meta_ in the prototype above) represents t
 
+** _meta.keyspace (type: Object)_
++
+The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. This parameter is only required if you are using a a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```
```suggestion
The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. 
This parameter is only required if you are using a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```

RayOffiah commented on modules/eventing/pages/eventing-advanced-keyspace-accessors.adoc:
```diff
@@ -516,6 +537,10 @@ The positional parameter (denoted by _meta_ in the prototype above) represents t
 
+** _meta.keyspace (type: Object)_
++
+The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. This parameter is only required if you are using a a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```
```suggestion
The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. 
This parameter is only required if you are using  a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```

As a general comment, it's usually a good idea to have repeated text set up as an Asciidoc attribute, instead of copying and pasting.

RayOffiah commented on modules/eventing/pages/eventing-advanced-keyspace-accessors.adoc:
```diff
@@ -682,6 +707,10 @@ The positional parameter (denoted by _meta_ in the prototype above) represents t
 
+** _meta.keyspace (type: Object)_
++
+The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. This parameter is only required if you are using a a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```
```suggestion
The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. 
This parameter is only required if you are using a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```

RayOffiah commented on modules/eventing/pages/eventing-advanced-keyspace-accessors.adoc:
```diff
@@ -797,6 +826,10 @@ The positional parameter (denoted by _meta_ in the prototype above) represents t
 
+** _meta.keyspace (type: Object)_
++
+The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. This parameter is only required if you are using a a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```
```suggestion
The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. 
This parameter is only required if you are using a  _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```

RayOffiah commented on modules/eventing/pages/eventing-advanced-keyspace-accessors.adoc:
```diff
@@ -902,6 +935,10 @@ The positional parameter (denoted by _meta_ in the prototype above) represents t
 
+** _meta.keyspace (type: Object)_
++
+The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. This parameter is only required if you are using a a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```
```suggestion
The keyspace of the document to be used for the operation with the JavaScript format "keyspace":{"bucket_name":_string_,"scope_name":_string_,"collection_name":_string_}. 
This parameter is only required if you are using a _binding_ with a wildcard for the scope and/or the collection. [.status]#Couchbase Server 7.1.1#
```

Yes, all these repeated blocks have the same error, `a a _binding_,`, so it's definitely worth using an attribute or an included file for the common text if possible. (That way the error only needs correcting in one place).

RayOffiah commented on modules/eventing/pages/eventing-handler-multiCollectionEventing.adoc:
```diff
@@ -0,0 +1,160 @@
+= Function: Multi Collection Eventing
+:description: pass:q[Show how to access the Data Service when Eventing is listening to multiple collections.]
+:page-edition: Enterprise Edition
+:tabs:
```
No real need for this attribute.

jon-strabala replied:
It's just how my other scriptlets were done, I probably use it as it bolds things.

RayOffiah commented on modules/eventing/pages/eventing-handler-multiCollectionEventing.adoc:
```diff
@@ -0,0 +1,160 @@
+* Will operate on three test documents.  Add them one at a time after the function is deployed.
+* Highlights the use of _meta.keyspace_
+
+[{tabs}] 
```
```suggestion
[tabs] 
```

RayOffiah commented on modules/eventing/pages/eventing-handler-multiCollectionEventing.adoc:
```diff
@@ -0,0 +1,160 @@
+multiCollectionEventing::
++
+--
+----
```
```suggestion
[source, javascript]
----
```

RayOffiah commented on modules/eventing/pages/eventing-lifecycle.adoc:
```diff
@@ -80,7 +80,8 @@ Typically you should set Function Scope to the bucket.scope that holds the colle
 | Listen To Location (the source keyspace)
-| The name of a collection currently defined on the cluster.
+| The name of a collection currently defined on the cluster. 
+As of 7.1.1 Eventing Functions can listen to multiple collections via a wildcard of "+++*+++" for the scope and/or the collection.
```
```suggestion
As of 7.1.1, Eventing Functions can listen to multiple collections via a wildcard of "+++*+++" for the scope and/or the collection.
```

RayOffiah reviewed (changes requested):
A couple of general comments:

- Filenames must always be lower case, with a hypen to separate words. We have occasionally run into problems with mixed case file names when compiling between Windows and Unix-like operating systems: https://docs.couchbase.com/home/contribute/pages.html#file-format-and-name
- Always handy to put common blocks of text in attribute or an included file. 
- For stuff like variables or wildcards, use backticks rather than quotes.
- One sentence per line.

jon-strabala committed:
Update modules/eventing/pages/eventing-Terminologies.adoc

Co-authored-by: Ray <77050471+RayOffiah@users.noreply.github.com>

jon-strabala committed:
TEST

jon-strabala committed:
Delete modules/eventing/partial directory

jon-strabala committed:
TEST

jon-strabala committed:
TEST

jon-strabala committed:
Delete keyspace.adoc

jon-strabala committed:
undo TEST and addback the PR reviews

jon-strabala committed:
Clarity

Clarity on why Basic Accessors do not work if wild carding a bucket binding.

jon-strabala committed:
Clarity

introduce the requirement of meta.keyspace if using a multi collection wildcard binding.

jon-strabala committed:
Add some italics
